reactor-netty-http/src/main/java/reactor/netty/http/server/HttpServerOperations.java

```diff
@@ -814,12 +814,24 @@ else if (msg instanceof HttpRequest) {
 				}
 				else {
 					request.release();
+					// HTTP/1.1 TLS Upgrade (RFC-2817) on empty requests (GET/HEAD/OPTIONS)
+					if (!isHttp2() && request.headers().contains(HttpHeaderNames.UPGRADE)) {
+						//force auto read to enable more accurate close selection now inbound is done
+						channel().config().setAutoRead(true);
+						onInboundComplete();
+					}
 				}
 				if (isHttp2()) {
 					//force auto read to enable more accurate close selection now inbound is done
 					channel().config().setAutoRead(true);
 					onInboundComplete();
 				}
+				else if (!isHttp2() && request.headers().contains(HttpHeaderNames.UPGRADE)) {
+					// HTTP/1.1 TLS Upgrade (RFC-2817) requests (GET/HEAD/OPTIONS) with empty / non-empty payload
+					// No need to call onInboundComplete(), it will be triggered by onInboundNext(...)
+					// since this is the last HTTP content
+					stopReadTimeout();
```

Comment:
The current logic doesn't invoke `onInboundComplete` if the request is not HTTP/2, because the only other expected use case was `websocket`. For `websocket` use case we should not invoke `onInboundComplete`.

Now with this commit, we want to change the logic to handle also aggregated HTTP/1.1 requests. For them we need to guarantee that `onInboundComplete`, which marks end of request.
With your current test case you don't see the issue, because the test itself invokes `complete`.
If you remove from `EchoAction`

```
if (message instanceof LastHttpContent) {
    emitter.complete();
}
```

You will see that completion is never invoked. Reactor Netty should guarantee that completion is triggered and not the user's logic.